tests/code_update.py

```diff
@@ -52,6 +57,126 @@ def test_verify_quotes(network, args):
+
+
+@reqs.description(
+    "Node with no security policy set but good digest matching ledger joins successfully"
+)
+@reqs.snp_only()
+def test_add_node_with_no_raw_security_policy_match_ledger(network, args):
+
+    LOG.info(
+        f"Change the entry for trusted security policies to not include a raw policy"
+    )
+    primary, _ = network.find_nodes()
+    network.consortium.retire_security_policy(
+        primary, DEFAULT_SNP_SECURITY_POLICY_DIGEST
+    )
+    network.consortium.add_new_security_policy(
+        primary, "", DEFAULT_SNP_SECURITY_POLICY_DIGEST
+    )
+
+    new_node = network.create_node("local://localhost")
+    network.join_node(new_node, args.package, args, timeout=3, env=dict())
+    network.trust_node(new_node, args)
+
+    LOG.info(f"Checking node joined with a blank security policy")
+    # TODO: Do this
+
+    network.consortium.retire_security_policy(
+        primary, DEFAULT_SNP_SECURITY_POLICY_DIGEST
+    )
+    network.consortium.add_new_security_policy(
+        primary, DEFAULT_SNP_SECURITY_POLICY, DEFAULT_SNP_SECURITY_POLICY_DIGEST
+    )
+
+
+@reqs.description(
+    "Node with no security policy set but good digest not matching ledger joins successfully"
+)
+@reqs.snp_only()
+def test_add_node_with_no_raw_security_policy_not_matching_ledger(network, args):
+
+    new_node = network.create_node("local://localhost")
+    network.join_node(new_node, args.package, args, timeout=3, env=dict())
+    network.trust_node(new_node, args)
+
+    LOG.info(f"Checking node joined with a blank security policy")
+    # TODO: Do this
+
+
+@reqs.description("Node where raw security policy doesn't match digest fails to join")
+@reqs.snp_only()
+def test_add_node_with_mismatched_security_policy_digest(network, args):
+
+    primary, others = network.find_nodes()
+    prev_nodes = [primary, *others]
+    try:
+        new_node = network.create_node("local://localhost")
+        network.join_node(
+            new_node,
+            args.package,
+            args,
+            timeout=3,
+            env={"SECURITY_POLICY": b64encode(b"invalid_security_policy").decode()},
+        )
+        network.trust_node(new_node, args)
+    except Exception:
+        ...
+
+    primary, others = network.find_nodes()
+    nodes = [primary, *others]
+    assert len(nodes) == len(prev_nodes), "Node joining unexpectedly succeeded"
+
+
+@reqs.description("Node with bad security policy digest fails to join")
+@reqs.snp_only()
+def test_add_node_with_bad_security_policy_digest(network, args):
+
+    primary, others = network.find_nodes()
+    prev_nodes = [primary, *others]
+
+    LOG.info(
+        f"Removing security policy set by node 0 so that a new joiner is seen as an unmatching policy"
+    )
+    network.consortium.retire_security_policy(
+        primary, DEFAULT_SNP_SECURITY_POLICY_DIGEST
+    )
+
+    new_node = network.create_node("local://localhost")
+    try:
+        network.join_node(new_node, args.package, args, timeout=3)
```

Comment:
Here and for other negative tests, could you `new_node.stop()` at the end of the test to make sure we don't have too many nodes running by the end of the suite?